File: factorial.py
#finds factorial of large numbers like(100,200,500,1000,2000...)

def findprod(p,q):
    
    #multiplies the digits of array with a number
    
    l=[]
    carry=0
    for i in range(0,len(p)):
        m=int(p[i])*q
        d=carry+m                                   #finding product, storing the digit at unit place in array, storing the carry 
        l.append(d%10)
        carry=(d-(d%10))//10
    l.append(carry)
    return l
        
def fact(n):

    #finds factorial by multiplying the digits stored in the array with next number
    
    res=['1']                                       #initializing array to 1
    for i in range(2,n+1):
        p=findprod(res,i)                           #multiplying the array with next number
        res=p
    return res
        
def main():
    
    #driver function
    
    n=int(input())                                      #takes input (the number whose factorial is to be calculated
    z=fact(n)                                           #calls fact() 
    for i in range(len(z)-1,-1,-1):                     #prints result
        print(z[i],end='')
    
main()

File: factorial.cpp
// program to find the factorial of large numbers
//can calculate upto 2000!
//for calculating higher numbers, increase the size of MAX
#include<bits/stdc++.h>
#define MAX 6000
using namespace std;

int prod(int *a,int n,int size)
{
// calculating the product of digits in array with a number.
	int i,carry=0,product=0;                                        				 //initializing carry and product to 0
	for(i=0;i<size;i++)
	{
		product=carry+(*(a+i))*n;                                                    //calculating the product
		*(a+i)=product%10;                                                           //storing the unit place digit in array
		carry=product/10;                                                            //storing carry
	}
	while(carry!=0)
	{
		*(a+i)=carry%10;                                                              //appending remaining carry to array
		carry=carry/10;
		i=i+1;
		size++;
	}
	return size;                                                                       //returning size of the changed array
}
void fact(int n)
{
//calculates the factorial by storing digits in array and multiplying them with next number.
	int a[MAX]={1};                                               				 	//intializing array to 1
	int i,size=1;                                                 					//size stores the size of array
	for(i=2;i<=n;i++)
	{
		size=prod(a,i,size);                                         				//mutliplying digits with next number
	}
	for(i=size-1;i>=0;i--)                                           				//prints the result and number of digits in result
	cout<<a[i];
	cout<<endl<<size;
	
}
int main()
{
//driver function
	int number;
	cout<<"enter the number:"<<endl;                                           
	cin>>number;                                                                    //takes input
	fact(number);                                                                   //calls fact() function to find factorial
	return 0;
}
